design/upgrade.md

```diff
@@ -0,0 +1,165 @@
+The cluster should be maintained in a healthy state the entire time.
+* **Migrations:** Breaking changes, as well as schema and data format changes should be handled through an automated migration processes.
+* **Rollback:** In the event that the upgrade is not successful, the Rook software should be rolled back to the previous version and cluster health should be restored.
+
+## **User Guide**
+Until automated upgrade support is available in Rook, we have authored a user guide that walks you through the steps to upgrade the software in a Rook cluster.
+Consideration is also provided in the guide for how to verify the cluster remains healthy during and after the upgrade process.
+Please refer to the [Rook Upgrade User Guide](../Documentation/upgrade.md) to learn more about the current Rook upgrade process.
+
+## **Detailed Design**
+The responsibility for performing and orchestrating an upgrade will be handled by an upgrade controller that runs as part of the Rook operator, in the same pod and process (similar to how the Rook volume provisioner is run).
+This controller will be responsible for carrying out the sequence of steps for updating each individual Rook component.
+Additionally, the controller will monitor cluster and component health during the upgrade process, taking corrective steps to restore health, up to and including a full rollback to the old version.
+
+### **Prerequisites**
+In order for the upgrade controller to begin an upgrade process, the following conditions must be met:
+* The cluster should be in a healthy state in accordance with our defined [health verification checks](#upgrade-health-verification).
+The upgrade controller should not begin an upgrade if the cluster is currently unhealthy.
+* Metadata for pods must be persistent.  If config files and other metadata only resides on an ephemeral empty dir for the pods (i.e., `dataDirHostPath` is not set), then the upgrade controller will not perform an upgrade.
+
+### **General Sequence**
+This section describes in a broad sense the general sequence of steps for upgrading a Rook cluster.
+Note that this sequence is modeled after the [Rook Upgrade User Guide](../Documentation/upgrade.md), including the cluster health checks described in the [health verification section](../Documentation/upgrade.md#health-verification).
+1. A new Rook software version is released, e.g. `v0.6.1`.
+1. The cluster administrator updates their Rook Cluster CRD to update the `versionTag` field using a command similar to `kubectl -n rook edit cluster.rook.io rook`
+1. The Rook operator receives the `update` event for the Cluster CRD instance and sees that the `versionTag` field has changed from its previous value.
+1. The upgrade controller begins a reconciliation to bring the clusters actual version value in agreement with the desired version from the cluster CRD.
+1. **Operator:** The operator pod itself is upgraded first since it is the host of the upgrade controller.
```

Comment:
ah, here's where you update the operator. If this is done first as suggested above, it simplifies the health checks here. Also if you had multiple clusters, you would only update the operator once anyway.

Reply:
yep, i am refactoring to make the Rook system namespace first (and a one time operation), then many clusters can be upgraded after that.